File: src/main/java/com/snackgame/server/member/domain/MemberRepository.java
package com.snackgame.server.member.domain;

import java.util.List;
import java.util.Optional;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import com.snackgame.server.member.exception.MemberNotFoundException;

@Repository
public interface MemberRepository extends JpaRepository<Member, Long> {

    default Member getById(Long id) {
        return findById(id)
                .orElseThrow(MemberNotFoundException::new);
    }

    @Query("select g from Guest g where g.id = :id")
    Optional<Guest> findGuestById(Long id);

    Optional<Member> findByName(Name name);

    default Member getByName(Name name) {
        return findByName(name)
                .orElseThrow(MemberNotFoundException::new);
    }

    boolean existsByName(Name name);

    List<Member> findByNameStringStartingWith(String prefix);

    Optional<SocialMember> findByProviderAndProvidedId(String provider, String providedId);
}

File: src/main/java/com/snackgame/server/member/domain/MemberRepository.kt
package com.snackgame.server.member.domain

import com.snackgame.server.member.exception.MemberNotFoundException
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface MemberRepository : JpaRepository<Member, Long?> {

    @Query("select g from Guest g where g.id = :id")
    fun findGuestById(id: Long): Guest?

    fun findByName(name: Name): Member?

    fun existsByName(name: Name?): Boolean

    fun findByNameStringStartingWith(prefix: String?): List<Member>

    fun findByProviderAndProvidedId(provider: String, providedId: String): Optional<SocialMember>
}

fun MemberRepository.getBy(id: Long): Member =
    findByIdOrNull(id) ?: throw MemberNotFoundException()

fun MemberRepository.getBy(name: Name): Member =
    findByName(name) ?: throw MemberNotFoundException()

